crates/fluvio-compression/src/lib.rs

```diff
@@ -90,9 +102,88 @@ impl std::fmt::Display for Compression {
     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
         match *self {
             Compression::None => write!(f, "none"),
-            Compression::Gzip => write!(f, "gzip"),
+            Compression::Gzip(_) => write!(f, "gzip"),
             Compression::Snappy => write!(f, "snappy"),
             Compression::Lz4 => write!(f, "lz4"),
         }
     }
 }
+
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[repr(i8)]
+pub enum GzipLevel {
+    Level1 = 1,
+    Level2 = 2,
+    Level3 = 3,
+    Level4 = 4,
+    Level5 = 5,
+    Level6 = 6,
+    Level7 = 7,
+    Level8 = 8,
+    Level9 = 9,
+}
+
+impl FromStr for GzipLevel {
+    type Err = CompressionError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use GzipLevel::*;
+        match s {
+            "1" => Ok(Level1),
+            "2" => Ok(Level2),
+            "3" => Ok(Level3),
+            "4" => Ok(Level4),
+            "5" => Ok(Level5),
+            "6" => Ok(Level6),
+            "7" => Ok(Level7),
+            "8" => Ok(Level8),
+            "9" => Ok(Level9),
+            _ => Err(CompressionError::UnknownGzipLevel(s.into())),
+        }
+    }
+}
+
+impl Default for GzipLevel {
+    fn default() -> Self {
+        GzipLevel::Level6
+    }
+}
+
+impl TryFrom<i8> for GzipLevel {
+    type Error = CompressionError;
+
+    fn try_from(v: i8) -> Result<Self, CompressionError> {
+        use GzipLevel::*;
+        match v {
+            1 => Ok(Level1),
+            2 => Ok(Level2),
+            3 => Ok(Level3),
+            4 => Ok(Level4),
+            5 => Ok(Level5),
+            6 => Ok(Level6),
+            7 => Ok(Level7),
+            8 => Ok(Level8),
+            9 => Ok(Level9),
+            _ => Err(CompressionError::UnknownCompressionFormat(format!(
+                "i8 representation: {}",
+                v
+            ))),
+        }
+    }
+}
+
+impl From<GzipLevel> for flate2::Compression {
+    fn from(level: GzipLevel) -> Self {
+        let int_level = level as u32;
+        if int_level < 1 || int_level > 9 {
+            panic!("Invalid GzipLevel discriminant: {int_level}")
```

Comment:
It is not safe to panic here. If someone forgets and adds `level10` it can kill the process in runtime. Better to cover all variants in `match`, and the correspondence will be controlled during compilation time.